dawn/prototype/AtlasFromNetcdf.cpp

```diff
@@ -0,0 +1,298 @@
+#include <atlas/mesh/HybridElements.h>
+#include <atlas/mesh/Nodes.h>
+#include <atlas/mesh/actions/BuildCellCentres.h>
+#include <atlas/mesh/actions/BuildEdges.h>
+#include <atlas/mesh/actions/BuildPeriodicBoundaries.h>
+#include <atlas/meshgenerator/detail/MeshGeneratorImpl.h>
+#include <atlas/output/Gmsh.h>
+#include <atlas/util/CoordinateEnums.h>
+
+static const int myPart = 0;
+
+template <typename loadT>
+static std::vector<loadT> LoadField(const netCDF::NcFile& dataFile, const std::string& name) {
+  netCDF::NcVar data = dataFile.getVar(name.c_str());
+  if(data.isNull()) {
+    return {};
+  }
+  assert(data.getDimCount() == 1);
+  int size = data.getDim(0).getSize();
+  std::vector<loadT> ret(size);
+  data.getVar(ret.data());
+  return ret;
+}
+
+template <typename loadT>
+static std::tuple<std::vector<loadT>, int, int> Load2DField(const netCDF::NcFile& dataFile,
+                                                            const std::string& name) {
+  netCDF::NcVar data = dataFile.getVar(name.c_str());
+  if(data.isNull()) {
+    return {};
+  }
+  assert(data.getDimCount() == 2);
+  int stride = data.getDim(0).getSize();
+  int elPerStride = data.getDim(1).getSize();
+  int size = stride * elPerStride;
+  std::vector<loadT> ret(size);
+  data.getVar(ret.data());
+  return {ret, stride, elPerStride};
+}
+
+static bool NodesFromNetCDF(const netCDF::NcFile& dataFile, atlas::Mesh& mesh) {
+  auto lon = LoadField<double>(dataFile, "vlon");
+  auto lat = LoadField<double>(dataFile, "vlat");
+  if(lon.size() == 0 || lat.size() == 0) {
+    std::cout << "lat / long variable not found\n";
+    return false;
+  }
+  if(lon.size() != lat.size()) {
+    std::cout << "lat / long not of consistent sizes!\n";
+    return false;
+  }
+
+  int numNodes = lat.size();
+
+  // define nodes and associated properties for Atlas meshs
+  mesh.nodes().resize(numNodes);
+  atlas::mesh::Nodes& nodes = mesh.nodes();
+  auto xy = atlas::array::make_view<double, 2>(nodes.xy());
+  auto lonlat = atlas::array::make_view<double, 2>(nodes.lonlat());
+
+  // we currently don't care about parts, so myPart is always 0 and remotde_idx == glb_idx
+  auto glb_idx_node = atlas::array::make_view<atlas::gidx_t, 1>(nodes.global_index());
+  auto remote_idx = atlas::array::make_indexview<atlas::idx_t, 1>(nodes.remote_index());
+  auto part = atlas::array::make_view<int, 1>(nodes.partition());
+
+  // no ghosts currently (ghost = false always) and no flags are set
+  auto ghost = atlas::array::make_view<int, 1>(nodes.ghost());
+  auto flags = atlas::array::make_view<int, 1>(nodes.flags());
+
+  auto radToLat = [](double rad) { return rad / (0.5 * M_PI) * 90; };
+  auto radToLon = [](double rad) { return rad / (M_PI)*180; };
+
+  for(int nodeIdx = 0; nodeIdx < numNodes; nodeIdx++) {
+    xy(nodeIdx, atlas::LON) = radToLon(lon[nodeIdx]);
+    xy(nodeIdx, atlas::LAT) = radToLat(lat[nodeIdx]);
+
+    lonlat(nodeIdx, atlas::LON) = xy(nodeIdx, atlas::LON);
```

Comment:
`lonlat` and `xy` are now set to the same value. Please comment why this makes sense, if it makes sense.

Reply:
On first sight it seemed that this is the default behavior of the `RegularMeshGenerator.cc` of atlas. Studying the code again this is in fact not the case, but the x,y index on the regular lattice is assigned. There is no such lattice for the ICON netcdf files, and to compute x,y values any of the map projections would need to be used. Hence, I chose to simply not assign `xy` and documented this behavior.